dbms/src/Storages/StorageDistributed.cpp

```diff
@@ -312,10 +312,19 @@ BlockInputStreams StorageDistributed::read(
 
     if (settings.optimize_skip_unused_shards)
     {
-        auto smaller_cluster = skipUnusedShards(cluster, query_info);
+        if (has_sharding_key)
+        {
+            auto smaller_cluster = skipUnusedShards(cluster, query_info);
 
-        if (smaller_cluster)
-            cluster = smaller_cluster;
+            if (smaller_cluster)
+                cluster = smaller_cluster;
+        }
+        else
+        {
+            LOG_WARNING(log,
```

Comment:
I don't think there is a need for warning, since the query may include multiple tables - with and without sharding key.

Reply:
The setting may be global as well ... A good point! Thanks! And yet we need to inform user in some other manner. I'll think of a better place to do this